shared/middleware/helpers/bucket_helper.rb

```diff
@@ -271,8 +271,31 @@ def restore_previous_version(encrypted_channel_id, filename, version_id, user_id
     owner_id, channel_id = storage_decrypt_channel_id(encrypted_channel_id)
     key = s3_path owner_id, channel_id, filename
 
-    response = s3.copy_object(bucket: @bucket, key: key, copy_source: "#{@bucket}/#{key}?versionId=#{version_id}")
-
+    begin
+      response = s3.copy_object(
+        bucket: @bucket,
+        key: key,
+        copy_source: "#{@bucket}/#{key}?versionId=#{version_id}"
+      )
+    rescue
+      response = s3.copy_object(
+        bucket: @bucket,
+        key: key,
+        copy_source: "#{@bucket}/#{key}",
+        metadata_directive: 'REPLACE',
+        metadata: {
+          abuse_score: get_abuse_score(encrypted_channel_id, filename).to_s,
+          failed_restore_at: Time.now.to_s,
+          failed_restore_from_version: version_id
+        }
+      )
+      Honeybadger.notify(
+        "Restore at Specified Version Failed. Restored most recent.",
```

Comment:
Do we maybe want to skip copying the latest version if the latest version is a delete marker?

Reply:
Great idea (and great catch), we're implementing and testing that.